NTT/fieldparameters.txt
The 62-bit prime is= 3073700804129980417
 1048576th (2^20th) root of unity is=  1414118249734601779 (mod 3073700804129980417)
14348907th (3^15th) root of unity is=   308414859194273485 (mod 3073700804129980417)

///////////////////////////// Roots related to powers of 2 \\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\

 524288th (2^19th) root of unity is=  821092576347561346 (mod 3073700804129980417)
 262144th (2^18th) root of unity is=  722168054604968797 (mod 3073700804129980417)
 131072th (2^17th) root of unity is=  2342921527381056840 (mod 3073700804129980417)
 65536th (2^16th) root of unity is=  2895001145261225276 (mod 3073700804129980417)
 32768th (2^15th) root of unity is=  185228267370228670 (mod 3073700804129980417)
 16384th (2^14th) root of unity is=  166706197685449053 (mod 3073700804129980417)
 8192th (2^13th) root of unity is=  1137714900568601731 (mod 3073700804129980417)
 4096th (2^12th) root of unity is=  356116677472641586 (mod 3073700804129980417)
 2048th (2^11th) root of unity is=  1288510276932018321 (mod 3073700804129980417)
 1024th (2^10th) root of unity is=  533143422376310736 (mod 3073700804129980417)
 512th (2^9th) root of unity is=  231059600000505185 (mod 3073700804129980417)
 256th (2^8th) root of unity is=  2084132838470242215 (mod 3073700804129980417)
 128th (2^7th) root of unity is=  667632013081810335 (mod 3073700804129980417)
 64th (2^6th) root of unity is=  2462566883598574694 (mod 3073700804129980417)
 32th (2^5th) root of unity is=  700305079945433357 (mod 3073700804129980417)
 16th (2^4th) root of unity is=  437380319823159113 (mod 3073700804129980417)
 8th (2^3th) root of unity is=  414345200490731620 (mod 3073700804129980417)
 4th (2^2th) root of unity is=  48123647995213473 (mod 3073700804129980417)
 2th (2^1th) root of unity is=  3073700804129980416 (mod 3073700804129980417)
 
///////////////////////////// Roots related to powers of 3 \\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\
 
 4782969th (3^14th) root of unity is=  2887552414139430178 (mod 3073700804129980417)
 1594323th (3^13th) root of unity is=  2589561880801976545 (mod 3073700804129980417)
 531441th (3^12th) root of unity is=  2055672124361686519 (mod 3073700804129980417)
 177147th (3^11th) root of unity is=  2867893238934135112 (mod 3073700804129980417)
 59049th (3^10th) root of unity is=  801543298049938745 (mod 3073700804129980417)
 19683th (3^9th) root of unity is=  2973529386614196309 (mod 3073700804129980417)
 6561th (3^8th) root of unity is=  1506635237690773038 (mod 3073700804129980417)
 2187th (3^7th) root of unity is=  1937907544692674569 (mod 3073700804129980417)
 729th (3^6th) root of unity is=  64543017631340700 (mod 3073700804129980417)
 243th (3^5th) root of unity is=  1884064513568972269 (mod 3073700804129980417)
 81th (3^4th) root of unity is=  1188866194714075022 (mod 3073700804129980417)
 27th (3^3th) root of unity is=  1697820560572790570 (mod 3073700804129980417)
 9th (3^2th) root of unity is=  2538470489641169034 (mod 3073700804129980417)
 3th (3^1th) root of unity is=  2632356178919102469 (mod 3073700804129980417)


NTT/parameters.cpp
// To compile: g++ -g -O0 -lm -lgmp -lntl  parameters.cpp 
// The main function receives two arguments argv[2]. 
// argv[1] is the vector size
// argv[2] is the number of the users
#include <NTL/BasicThreadPool.h>
#include <NTL/vector.h>
#include <NTL/ZZ.h>
#include <NTL/ZZ_pXFactoring.h>


using namespace std;
using namespace NTL;

using NTL::ZZ;

using NTL::Vec;

// Returns 1 if Number is a power of two greater than 1.
bool isPowerOfTwo(long Number){
    long n=2;
    while (1){
        if (n > Number)
            return 0;
        else if (n == Number)
            return 1;
        n *= 2;
    }
    return 0;
}

long whichRoot2(long Number){

    if (!isPowerOfTwo(Number)){
    cout<< Number << " is not a power of 2 or is 1"<<endl;
    return 0;
    }
    long n=2;
    long power=1;
    while (1){
        if (n == Number)
            return power;
        power += 1;    
        n *= 2;
    }
    return 0;
}

long whichRoot3(long Number){


    long n=3;
    long power=1;
    while (1){
        //TODO: check if equality should be considered
        if (n >= Number)
            return power;
        power += 1;    
        n *= 3;
    }
    return 0;
}

int main(int argc, char* argv[])
{
    if (argc < 3)
    {
        cout <<argc << "is too few for arguments"<<endl;
        cout <<"usage: " << "Vector_length Number_Users "<<endl;
        return 0;
    }
    
   ZZ theprime(3073700804129980417);       
   ZZ_p::init(theprime);
   ZZ_p rootTwo(1414118249734601779);
   Vec<ZZ_p> allRootTwo;
   allRootTwo.SetLength(21);
   allRootTwo[0] = 1;
   allRootTwo[20] = rootTwo;
   ZZ_p rootThree(308414859194273485);
   Vec<ZZ_p> allRootThree;
   allRootThree.SetLength(16);
   allRootThree[0] = 1;
   allRootThree[15] = rootThree;
   ZZ_p nth2(1048576);
   ZZ_p nth3(14348907);

   cout <<"The 62-bit prime is= " << theprime<< endl;
   cout <<" "<<nth2<<"th (2^20th) root of unity is=  " << rootTwo <<" (mod "<<theprime<<")" << endl;
   cout <<nth3<<"th (3^15th) root of unity is=   " << rootThree<<" (mod "<<theprime<<")" << endl;
  
   ZZ_p twos = rootTwo;
   long k=2;
   for (long i = 1 ; i < 20 ; i++ ){
       twos=twos*twos;
        allRootTwo[20-i] = twos;
       k=k*2;
   }
   ZZ_p threes = rootThree;
   k=3;
    for (long i = 1 ; i < 15 ; i++ ){
       threes=threes*threes*threes;
       allRootThree[15-i] = threes;
       k=k*3;
   }

   long vectorSize = atoi(argv[1]);
   long numUsers = atoi(argv[2]);
   long degree = numUsers/3;
   long numCrPrt = numUsers/6;
    cout << "vectorSize = "<< vectorSize << endl;
    cout << "numUsers = "<< numUsers << endl;
   if(!isPowerOfTwo(degree))
   {
       cout <<"degree before reduction= "<< degree <<endl;
    long degtmp= degree>>1;
    degree = 1 ;
    while(degtmp){
        degtmp>>=1;
        degree *= 2;
        
    }
   }

    cout << "degree = "<< degree <<"   (After reduction)"<< endl;
    cout << "numCrPrt = "<< numCrPrt << endl;
    cout << "block length = "<< degree - numCrPrt << endl;
    long whichr2 = whichRoot2(degree);
    long whichr3 = whichRoot3(degree);

    cout << allRootTwo[whichr2] << " is 2^"<<whichr2<<"th root of unity"<<endl;
    cout << allRootThree[whichr3] << " is 3^"<<whichr3<<"th root of unity"<<endl; 
 
	return 0;
	
}
